sdklab/regressions/regression_github_990.py

```diff
@@ -0,0 +1,284 @@
+        """
+        nonlocal fail_sock_send_until_reconnect
+
+        if fail_sock_send_until_reconnect:
+            log_func("-----------DROPPING {} bytes".format(len(buf)))
+            return len(buf)
+        else:
+            count = old_sock_send(buf)
+            log_func("---------- SENT {} bytes".format(count))
+            return count
+
+    @functools.wraps(old_sock_close)
+    def new_sock_close():
+        """
+        `sock_close` needs to restore normal operation, thus removing the failure conditions when
+        a new socket is opened.
+        """
+        nonlocal fail_sock_send_until_reconnect
+
+        if fail_sock_send_until_reconnect:
+            log_func("-----------RESTORING SOCKET behavior")
+            fail_sock_send_until_reconnect = False
+        return old_sock_close()
+
+    @functools.wraps(old_reconnect)
+    def new_reconnect(*args, **kwargs):
+        """
+        `reconnect` needs to raise an exception when we need it to in order to simulate the exception
+        that lead to the discovery of this bug.
+        """
+        nonlocal raise_on_next_reconnect
+        if raise_on_next_reconnect:
+            log_func(
+                "----------- RECONNECT CALLED. raising {}".format(
+                    type(exception_to_raise_on_reconnect)
+                )
+            )
+            raise_on_next_reconnect = False
+            raise exception_to_raise_on_reconnect
+        return old_reconnect(*args, **kwargs)
+
+    paho._sock_send = new_sock_send
+    paho._sock_recv = new_sock_recv
+    paho._sock_close = new_sock_close
+    paho.publish = new_publish
+    paho.reconnect = new_reconnect
+
+
+async def run_test(
+    connection_retry, auto_connect, send_should_succeed, exception_to_raise_on_reconnect
+):
+    try:
+        print("*" * 80)
+        print()
+        print(
+            "Running test with connection_retry={}, auto_connect={}, and exception_to_raies_on_reconnect={}".format(
+                connection_retry, auto_connect, type(exception_to_raise_on_reconnect)
+            )
+        )
+        print()
+        print("*" * 80)
+
+        # Create instance of the device client using the connection string
+        device_client = IoTHubDeviceClient.create_from_connection_string(
+            test_env.DEVICE_CONNECTION_STRING,
+            keep_alive=10,
+            connection_retry=connection_retry,
+            auto_connect=auto_connect,
+        )
+        logging_hook.hook_device_client(device_client)
+
+        hack_paho_to_disconnect_after_publish(device_client, exception_to_raise_on_reconnect)
+
+        # Connect the device client.
+        await device_client.connect()
+
+        try:
+            print("Sending message...")
+            await device_client.send_message("This is a message that is being sent")
+            print("Message successfully sent!")
+            assert send_should_succeed
+        except Exception as e:
```

Comment:
Again, let's try and use a custom defined exception here, this makes me very nervous about false positives